### PR TITLE
Combine Packet and PacketList into ReceivedMessage on receiving end

### PR DESCRIPTION
**Pull request by huffman**
No need to review yet. This is from a few weeks ago and I need to run back through it and do more testing.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2a416ebc61a0f7812071207501a0d3e2d3e6bc0d available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c6e4a58d9bbb790d33dd600fa5fa2e74f8ea5520 available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 07d21514c5587417b353ac8c71deddcf5ceb4164 available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dac491ef419bc3fd3052f1ffadfb3950f3b4b77c available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

File: libraries/networking/src/AssetClient.cpp [hunk 1]
```diff
@@ -256,8 +250,25 @@ void AssetClient::handleAssetGetReply(QSharedPointer<NLPacketList> packetList, S
...
+                connect(message.data(), &ReceivedMessage::progress, this, [this, length, message, callbacks]() {
+                    qDebug() << "Progress: " << message->getSize(), length;
+                    callbacks.progressCallback(message->getSize(), length);
+                });
+                connect(message.data(), &ReceivedMessage::completed, this, [this, message, error, callbacks]() {
+                    if (message->failed()) {
+                        qDebug() << "Failed to received asset :(";
+                        callbacks.completeCallback(false, AssetServerError::NoError, QByteArray());
+                    } else {
+                        qDebug() << "Succesfully received asset!";
```
Comment (huffman) [inline libraries/networking/src/AssetClient.cpp]: I will remove these

File: libraries/networking/src/AssetRequest.cpp [hunk 2]
```diff
@@ -105,12 +105,12 @@ void AssetRequest::start() {
...
                 
                 // we need to check the hash of the received data to make sure it matches what we expect
-                if (hashData(data).toHex() == _hash) {
+                if (hashData(data).toHex() == _hash || true) {
                     memcpy(_data.data() + start, data.constData(), data.size());
                     _totalReceived += data.size();
                     emit progress(_totalReceived, _info.size);
                     
-                    saveToCache(getUrl(), data);
+                    //saveToCache(getUrl(), data);
```
Comment (huffman) [inline libraries/networking/src/AssetRequest.cpp]: I need to revert this

**Comment by hifi-gustavo** [comment]
Build console output for commit: 669d3c3523afffe2074d3ea78f557964ed613108 available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by huffman** [comment]
Leaving this as do-not-merge until after the Thanksgiving weekend to ensure I'm available for any potential fallout.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 99dd9646bcc6b3a6aba7a4d211ff678c692d01b4 available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1b0413f8e3a6582510950ee493ed083b8e158a78 available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 304b105dac6025c8f8235d343da608b084d8a18b available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5849befb3d167b918b62071ca3e04657c220fd8c available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

File: assignment-client/src/Agent.cpp [hunk 3]
```diff
@@ -77,46 +77,46 @@ Agent::Agent(NLPacket& packet) :
...
+            int piggyBackedSizeWithHeader = message->getSize() - statsMessageLength;
+            
             auto buffer = std::unique_ptr<char[]>(new char[piggyBackedSizeWithHeader]);
-            memcpy(buffer.get(), packet->getPayload() + statsMessageLength, piggyBackedSizeWithHeader);
+            memcpy(buffer.get(), message->getRawMessage() + statsMessageLength, piggyBackedSizeWithHeader);
 
-            auto newPacket = NLPacket::fromReceivedPacket(std::move(buffer), piggyBackedSizeWithHeader, packet->getSenderSockAddr());
-            packet = QSharedPointer<NLPacket>(newPacket.release());
+            auto newPacket = NLPacket::fromReceivedPacket(std::move(buffer), piggyBackedSizeWithHeader, message->getSenderSockAddr());
+            message = QSharedPointer<ReceivedMessage>(new ReceivedMessage(*newPacket));
```
Comment (Atlante45) [inline assignment-client/src/Agent.cpp]: Use `QSharedPointer<ReceivedMessage>::create`

File: interface/src/octree/OctreePacketProcessor.cpp [hunk 4]
```diff
@@ -39,41 +39,43 @@ void OctreePacketProcessor::processPacket(QSharedPointer<NLPacket> packet, Share
...
             // construct a new packet from the piggybacked one
             auto buffer = std::unique_ptr<char[]>(new char[piggybackBytes]);
-            memcpy(buffer.get(), packet->getPayload() + statsMessageLength, piggybackBytes);
+            memcpy(buffer.get(), message->getRawMessage() + statsMessageLength, piggybackBytes);
+            qDebug() << "Got piggyback, read " << piggybackBytes << " bytes";
             
-            auto newPacket = NLPacket::fromReceivedPacket(std::move(buffer), piggybackBytes, packet->getSenderSockAddr());
-            packet = QSharedPointer<NLPacket>(newPacket.release());
+            auto newPacket = NLPacket::fromReceivedPacket(std::move(buffer), piggybackBytes, message->getSenderSockAddr());
+            message = QSharedPointer<ReceivedMessage>(new ReceivedMessage(*newPacket.release()));
```
Comment (Atlante45) [inline interface/src/octree/OctreePacketProcessor.cpp]: Use `QSharedPointer<ReceivedMessage>::create`

File: libraries/networking/src/udt/PacketList.cpp [hunk 5]
```diff
@@ -50,6 +50,10 @@ PacketList::PacketList(PacketList&& other) :
 {
 }
 
+HifiSockAddr PacketList::getSenderSockAddr() const {
+    return _packets.size() > 0 ? _packets.front().get()->getSenderSockAddr() : HifiSockAddr();
```
Comment (Atlante45) [inline libraries/networking/src/udt/PacketList.cpp]: Unnecessary `.get()`

File: tools/udt-test/src/UDTTest.cpp [hunk 6]
```diff
@@ -176,8 +176,9 @@ UDTTest::UDTTest(int& argc, char** argv) :
     } else {
         // this is a receiver - in case there are ordered packets (messages) being sent to us make sure that we handle them
         // so that they can be verified
-        _socket.setPacketListHandler(
-            [this](std::unique_ptr<udt::PacketList> packetList) { handlePacketList(std::move(packetList)); });
+        // TODO Fix support for message testing
+        //_socket.setMessageHandler(
+            //[this](std::unique_ptr<udt::PacketList> packetList) { handlePacketList(std::move(packetList)); });
```
Comment (Atlante45) [inline tools/udt-test/src/UDTTest.cpp]: Should we fix that before merge?

**Comment by hifi-gustavo** [comment]
Build console output for commit: ce882a9c112ea1eb91ae2a1a0713e60cf9a5f96d available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 89c412c3247f4b11892e1dde862abfcdbe2948d0 available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by huffman** [comment]
Status: Updated UDTTest, but running into a weird issue where the stats aren't showing up. Debugging.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 34138f07bd8e3fc28eea7f1d121b5c6f11c8bb54 available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dd6c8cbe335a4aaa511bc8add3ad9d1dd9cc10ae available at: https://gist.github.com/e90936eb2a619bf797b0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6431.dmg

**Pull request merged**